File: src/memory.py


class Memory:

    def __init__(self, size: int) -> None:
        self.size = size
        self.memory = [0] * size
    

    def get_data(self, address: int, size: int) -> list:
        """
        Get data from memory.
        """
        return self.memory[address:address + size]


    def store_data(self, address: int, data: list) -> None:
        """
        Store data to memory.
        """
        self.memory[address:address + len(data)] = data

File: c++/src/memory.cpp
#include "memory.hh"
#include <string.h>


using namespace memory;


Memory::Memory(size_t stackSize, size_t videoSize) {
    this->stackSize = stackSize;
    this->videoSize = videoSize;
    this->stack = new Byte[this->stackSize];
    this->video = new Byte[this->videoSize];
}


Memory::~Memory()
{
    delete[] this->stack;
    delete[] this->video;
}


void Memory::setByte(Address address, Byte data) {
    this->stack[address] = data;
}


void Memory::setBytes(Address address, const Byte* data, size_t size) {
    memcpy(this->stack + address, data, size);
}


Byte Memory::getByte(Address address) const {
    return this->stack[address];
}


const Byte* Memory::getBytes(Address address, size_t size) const {
    Byte* data = new Byte[size];
    memcpy(data, this->stack + address, size);
    return data;
}


Byte* Memory::getBytesMutable(Address address) {
    return this->stack + address;
}

